sycamore/docset.py

```diff
@@ -328,6 +328,47 @@ def merge(self, merger: ElementMerger, **kwargs) -> "DocSet":
         merged = Merge(self.plan, merger=merger, **kwargs)
         return DocSet(self.context, merged)
 
+    def regex_replace(self, spec: list[tuple[str, str]], **kwargs) -> "DocSet":
+        """
+        Performs regular expression replacement (using re.sub()) on the
+        text_representation of every Element in each Document.
+
+        Example:
+            .. code-block:: python
+            ds = context.read.binary(paths, binary_format="pdf")
+                .partition(partitioner=UnstructuredPdfPartitioner())
+                .regex_replace([
+                    (r"\d+", "1313"),
+                    (r"old", "new"),
+                ])
+                .explode()
+        """
+        from sycamore.transforms import RegexReplace
+
+        plan = RegexReplace(self.plan, spec, **kwargs)
+        return DocSet(self.context, plan)
+
+    def coalesce_whitespace(self, **kwargs) -> "DocSet":
+        """
+        Reduce long repetitions of whitespace characters to a single space
+        in the text_representation of every Element in each Document.
+
+        Example:
+            .. code-block:: python
+            ds = context.read.binary(paths, binary_format="pdf")
+                .partition(partitioner=UnstructuredPdfPartitioner())
+                .coalesce_whitespace()
+                .explode()
+        """
+        return self.regex_replace(
+            [
+                (r"\s+", " "),
+                (r"^ ", ""),
+                (r" $", ""),
+            ],
+            **kwargs,
+        )
+
```

Comment:
I think in the interest of not adding too much to `docset` we can elide this method? We can just do this in the script, right?

Reply:
We can.  I was thinking to provide ease-of-use, and self-documentation for common cases.  I think the growth of DocSet is inevitable and we should make a plan for how to organize it.  Or, maybe we realize that DocSet is just a doorway to transformer classes and we drop the habit of creating 1:1 helper methods for transformers.  I dunno.

Reply:
What if we make the transformer:
def regex_replace(self, res, **kwargs) -> "DocSet":
    # handle single re, list of res here.

and have
COALESCE_WHITESPACE = [ (r"\s+", " "), (r"^ ", ""), (r" $", "") ]

so people can call it as
docset
   .regex_replace(transforms.COALESCE_WHITESPACE)

Reply:
I also think a regex_replace/apply is a little more scalable than coalesce_whitespace. I'd even think having a defined coalesce function and having the user pass that in to .map is fine